c_src/notif_set_tlh_htable.c

```diff
@@ -0,0 +1,198 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <limits.h>
+#include "notif_set.h"
+#include "optics_htable/htable.h"
+#include "message.h"
+
+struct Notif_set {
+    // a two-level hash table where the value of each
+    // key points to another (level2) hash table 
+    struct htable lev1; // directory watch descriptors
+    size_t len;
+};
+
+struct Notif_set_itr {
+    struct Notif_set *tab;
+    struct htable_bucket *itr1, *itr2;
+};
+
+struct Notif_set *notif_set_new()
+{
+    struct Notif_set *tlh = malloc(sizeof(struct Notif_set));
+    if(tlh == NULL) {
+        return NULL;
+    }
+    htable_init(&tlh->lev1);
+    tlh->len = 0;
+    return tlh;
+}
+
+int notif_set_destroy(struct Notif_set *tab)
+{
+    struct htable_bucket *itr;
+    // destroy each level2 table
+    for(itr = htable_next(&tab->lev1, NULL); itr; itr = htable_next(&tab->lev1, itr)) {
+        htable_reset((struct htable*) itr->value);
+    }
+    htable_reset(&tab->lev1);
+    free(tab);
+
+    return 1;
+}
+
+int notif_set_insert(struct Notif_set *tab, struct Message *msg)
+{
+    // IMPORTANT: Each htable owns its keys, but the 
+    //            values are owned by the user.
+    if(msg == NULL) {
+        return 0;
+    }
+    struct Key *key = htable_key_int(msg->wd);
+    struct htable_ret get_ret = htable_get(&tab->lev1, key);
+    struct htable *lev2_tab = (struct htable*) get_ret.value;
+    if(!get_ret.ok) {
+        // We have a new directory wd; insert to level1
+        // First create a level2 table for this wd (the value)
+        lev2_tab = malloc(sizeof(struct htable));
+        assert(lev2_tab != NULL);
```

Comment:
Here you're asserting on the return value of the malloc call while in the `notif_set_new` function you're returning an error response instead. I being consistent in the error handling and in this case, returning an error instead of asserting.